perception/map_based_prediction/src/path_generator.cpp

```diff
@@ -0,0 +1,346 @@
+    return interpolated_path;
+  }
+
+  std::vector<double> base_path_x;
+  std::vector<double> base_path_y;
+  std::vector<double> base_path_z;
+  std::vector<double> base_path_s;
+  for (size_t i = 0; i < base_path.size(); ++i) {
+    base_path_x.push_back(base_path.at(i).position.x);
+    base_path_y.push_back(base_path.at(i).position.y);
+    base_path_z.push_back(base_path.at(i).position.z);
+    base_path_s.push_back(tier4_autoware_utils::calcSignedArcLength(base_path, 0, i));
+  }
+
+  std::vector<double> resampled_s(frenet_predicted_path.size());
+  for (size_t i = 0; i < frenet_predicted_path.size(); ++i) {
+    resampled_s.at(i) = frenet_predicted_path.at(i).s;
+  }
+
+  // Spline Interpolation
+  std::vector<double> slerp_ref_path_x =
+    interpolation::slerp(base_path_s, base_path_x, resampled_s);
+  std::vector<double> slerp_ref_path_y =
+    interpolation::slerp(base_path_s, base_path_y, resampled_s);
+  std::vector<double> slerp_ref_path_z =
+    interpolation::slerp(base_path_s, base_path_z, resampled_s);
+
+  interpolated_path.resize(interpolate_num);
+  for (size_t i = 0; i < interpolate_num - 1; ++i) {
+    geometry_msgs::msg::Pose interpolated_pose;
+    const double curr_x = slerp_ref_path_x.at(i);
+    const double curr_y = slerp_ref_path_y.at(i);
+    const double next_x = slerp_ref_path_x.at(i + 1);
+    const double next_y = slerp_ref_path_y.at(i + 1);
+    const double yaw = getPointYaw(curr_x, curr_y, next_x, next_y);
+    interpolated_pose.position = tier4_autoware_utils::createPoint(
+      slerp_ref_path_x.at(i), slerp_ref_path_y.at(i), slerp_ref_path_z.at(i));
+    interpolated_pose.orientation = tier4_autoware_utils::createQuaternionFromYaw(yaw);
+    interpolated_path.at(i) = interpolated_pose;
+  }
+  interpolated_path.back().position = tier4_autoware_utils::createPoint(
+    slerp_ref_path_x.back(), slerp_ref_path_y.back(), slerp_ref_path_z.back());
+  interpolated_path.back().orientation = interpolated_path.at(interpolate_num - 2).orientation;
+
+  return interpolated_path;
+}
+
+PredictedPath PathGenerator::convertPredictedPath(
+  const TrackedObject & object, const FrenetPath & frenet_predicted_path, const PosePath & ref_path)
+{
+  PredictedPath predicted_path;
+  predicted_path.time_step = rclcpp::Duration::from_seconds(sampling_time_interval_);
+  predicted_path.path.resize(ref_path.size());
+  for (size_t i = 0; i < predicted_path.path.size(); ++i) {
+    // Reference Point from interpolated reference path
+    const auto ref_pose = ref_path.at(i);
+    const auto ref_yaw = tf2::getYaw(ref_pose.orientation);
+
+    // Frenet Point from frenet predicted path
+    const auto frenet_point = frenet_predicted_path.at(i);
+
+    // Converted Pose
+    geometry_msgs::msg::Pose predicted_pose;
+    predicted_pose.position.x = ref_pose.position.x - std::sin(ref_yaw) * frenet_point.d;
+    predicted_pose.position.y = ref_pose.position.y + std::cos(ref_yaw) * frenet_point.d;
```

Comment:
Better to use `tier4_autoware_utils::calcOffsetPose`

Reply:
These are the coefficients for calculating the longitudianl and lateral position. Note that we use quintic and quartic interpolation for the calculation.

Reply:
done